pkg/controller/machine/controller.go

```diff
@@ -197,7 +197,7 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
 		// can be unlinked from a machine when the node goes NotReady and is removed
 		// by cloud controller manager. In that case some machines would never get
 		// deleted without a manual intervention.
-		if _, exists := m.ObjectMeta.Annotations[ExcludeNodeDrainingAnnotation]; !exists && m.Status.NodeRef != nil {
+		if _, exists := m.ObjectMeta.Annotations[ExcludeNodeDrainingAnnotation]; !exists && m.Status.NodeRef != nil && r.isNodeReady(ctx, m.Status.NodeRef.Name) {
```

Comment:
if anything I think we should probably check the node is unreachable, still not sure we want to hard delete